File: src/main/java/editor/gui/filter/editor/ManaCostFilterPanel.java
package editor.gui.filter.editor;

import java.awt.Color;

import javax.swing.BoxLayout;
import javax.swing.JTextField;
import javax.swing.event.DocumentEvent;

import editor.database.attributes.ManaCost;
import editor.filter.Filter;
import editor.filter.leaf.FilterLeaf;
import editor.filter.leaf.ManaCostFilter;
import editor.gui.generic.ComboBoxPanel;
import editor.gui.generic.DocumentChangeListener;
import editor.util.Containment;

/**
 * This class represents a panel corresponding to a filter that groups
 * cards by mana cost.
 *
 * @author Alec Roelke
 */
public class ManaCostFilterPanel extends FilterEditorPanel<ManaCostFilter>
{
    /** Color to display if the entered cost is valid (i.e. can be parsed according to {@link ManaCost#tryParseManaCost(String)}). */
    public static final Color VALID = Color.WHITE;
    /** Color to display if the entered cost is invalid (i.e. can't be parsed according to {@link ManaCost#tryParseManaCost(String)}). */
    public static final Color INVALID = Color.PINK;

    /**
     * Combo box indicating containment.
     */
    private ComboBoxPanel<Containment> contain;
    /**
     * Text field for entering the text version of the mana cost.
     */
    private JTextField cost;

    /**
     * Create a new ManaCostFilterPanel.
     */
    public ManaCostFilterPanel()
    {
        super();
        setLayout(new BoxLayout(this, BoxLayout.X_AXIS));

        contain = new ComboBoxPanel<>(Containment.values());
        add(contain);
        cost = new JTextField();
        cost.getDocument().addDocumentListener(new DocumentChangeListener()
        {
            @Override
            public void update(DocumentEvent e)
            {
                cost.setBackground(ManaCost.tryParseManaCost(cost.getText()).map((c) -> VALID).orElse(INVALID));
            }
        });
        add(cost);
    }

    /**
     * Create a new ManaCostFilterPanel using the given ManaCostFilter
     * to set the contents of its fields.
     *
     * @param f filter to use for initialization
     */
    public ManaCostFilterPanel(ManaCostFilter f)
    {
        this();
        setContents(f);
    }

    @Override
    public Filter filter()
    {
        ManaCostFilter filter = new ManaCostFilter();
        filter.contain = contain.getSelectedItem();
        filter.cost = ManaCost.tryParseManaCost(cost.getText()).orElse(new ManaCost());
        return filter;
    }

    @Override
    public void setContents(ManaCostFilter filter)
    {
        contain.setSelectedItem(filter.contain);
        cost.setText(filter.cost.toString());
    }

    @Override
    public void setContents(FilterLeaf<?> filter) throws IllegalArgumentException
    {
        if (filter instanceof ManaCostFilter f)
            setContents(f);
        else
            throw new IllegalArgumentException("Illegal mana cost filter " + filter.type());
    }
}

File: src/main/scala/editor/gui/filter/editor/ManaCostFilterPanel.scala
package editor.gui.filter.editor

import editor.database.attributes.CardAttribute
import editor.database.attributes.ManaCost
import editor.filter.leaf.FilterLeaf
import editor.filter.leaf.ManaCostFilter
import editor.gui.filter.FilterSelectorPanel
import editor.gui.generic.ComboBoxPanel
import editor.gui.generic.DocumentChangeListener
import editor.util.Containment

import java.awt.Color
import javax.swing.BoxLayout
import javax.swing.JTextField
import javax.swing.event.DocumentEvent

/**
 * Convenience constructors for [[ManaCostFilterPanel]].
 * @author Alec Roelke
 */
object ManaCostFilterPanel {
  /** @return a new, empty [[ManaCostFilterPanel]] */
  def apply(selector: FilterSelectorPanel) = new ManaCostFilterPanel(selector)

  /**
   * Create a new [[ManaCostFilterPanel]] pre-populated with the contents of a filter.
   * 
   * @param filter filter to use to populate the new panel
   * @return a new mana cost filter panel containing the string version of the filter's cost
   */
  def apply(filter: ManaCostFilter, selector: FilterSelectorPanel) = {
    val panel = new ManaCostFilterPanel(selector)
    panel.setContents(filter)
    panel
  }
}

/**
 * A panel for customizing a [[ManaCostFilter]] by setting the symbols in it and how it should contain them.
 * Symbols are set using text, with each of their strings surrounded by braces ({}).
 * 
 * @author Alec Roelke
 */
class ManaCostFilterPanel(selector: FilterSelectorPanel) extends FilterEditorPanel[ManaCostFilter] {
  private val Valid = Color.WHITE
  private val Invalid = Color.PINK

  setLayout(BoxLayout(this, BoxLayout.X_AXIS))
  
  private val contain = ComboBoxPanel(Containment.values)
  add(contain)
  private val cost = JTextField()
  cost.getDocument.addDocumentListener(new DocumentChangeListener {
    override def update(e: DocumentEvent) = cost.setBackground(ManaCost.parse(cost.getText).map(_ => Valid).getOrElse(Invalid))
  })
  add(cost)

  protected override val attribute = CardAttribute.ManaCost

  override def filter = {
    val filter = ManaCostFilter()
    filter.faces = selector.faces
    filter.contain = contain.getSelectedItem
    filter.cost = ManaCost.parse(cost.getText).getOrElse(ManaCost())
    filter
  }

  override def setFields(filter: ManaCostFilter) = {
    contain.setSelectedItem(filter.contain)
    cost.setText(filter.cost.toString)
  }
}
